Don't let automatic scaling destroy the sparsity pattern

lindsayad opened the pull request:
We call through to only kernels during the automatic scaling
calculation. These kernels do not capture the sparsity pattern
needed for things like DGKernels and InterfaceKernels, so if
we don't tell PETSc to not change the sparsity pattern, it will
shrink the allocation (I believe). Then when we actually call
through to compute the real Jacobian, we have to do a bunch of
mallocs in MatSetValues which makes the simulation super slow.
To fix this, we use @fdkong MatResetPreallocation function.

This also deletes FEProblemBase::computeResidualType which was
not being used anywhere.

Refs #12601

lindsayad committed:
Don't let automatic scaling destroy the sparsity pattern

We call through to only kernels during the automatic scaling
calculation. These kernels do not capture the sparsity pattern
needed for things like DGKernels and InterfaceKernels, so if
we don't tell PETSc to not change the sparsity pattern, it will
shrink the allocation (I believe). Then when we actually call
through to compute the real Jacobian, we have to do a bunch of
mallocs in MatSetValues which makes the simulation super slow.
To fix this, we use @fdkong MatResetPreallocation function.

This also deletes FEProblemBase::computeResidualType which was
not being used anywhere.

Refs #12601

fdkong commented on framework/include/problems/FEProblemBase.h:
```diff
@@ -1065,12 +1065,7 @@ class FEProblemBase : public SubProblem, public Restartable
-  virtual void computeResidualType(const NumericVector<Number> & soln,
-                                   NumericVector<Number> & residual,
-                                   TagID tag);
+
```
There was a reason to have this function at the first place, but I just forgot what it was

lindsayad replied:
`git grep` shows that it's not being used anywhere, at least not in the moose repository. I guess maybe apps might be calling it, which will show up in the testing

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3240,6 +3240,11 @@ NonlinearSystemBase::computeScalingJacobian(NonlinearImplicitSystem & sys)
       displaced_problem->systemBaseNonlinear().applyScalingFactors(inverse_scaling_factors);
+
+    // Now it's essential that we reset the sparsity pattern of the matrix
+    auto ierr = MatResetPreallocation(petsc_matrix.mat());
```
Do you mind to add something like: 

#if LIBMESH_HAVE_PETSC
#endif

fdkong replied:
You need to check PETSC version as well. The API is pretty new.

fdkong replied:
#if !PETSC_VERSION_LESS_THAN(3, 8, 0)
#endif

lindsayad replied:
I've already wrapped the entire contents of this function with `LIBMESH_HAVE_PETSC`. With respect to the version, do you think we should just not do scaling if we can't reset the preallocation? Otherwise the first real Jacobian evaluation will be exceedingly slow

fdkong replied:
Right, we should say something like: autoscaling is ignored because you are using too old PETSc. 

I do not like people complain about the performance because of their faults.

lindsayad replied:
👍

fdkong commented:
It it the right fix, I believe.

lindsayad committed:
Guard against old PETSc versions

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/360974/) on 24de027 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13670/site)

This comment will be updated on new commits.

lindsayad committed:
Restrict max_cpu to 2 for this suite of tests

PETSc appears not to allow restting of the preallocation on
processes which have no dofs

lindsayad commented on test/tests/kernels/simple_transient_diffusion/tests:
```diff
@@ -39,6 +39,7 @@
     requirement = "We shall be able to solve an initially poorly scaled problem by using MOOSE's automatic scaling feature in parallel"
     cli_args = 'Executioner/automatic_scaling=true -pc_type hypre -pc_hypre_type boomeramg'
     prereq = 'automatic-scaling-done-once'
+    max_parallel = 2 # Can't reset preallocation in PETSc when there are no dofs on a process
```
@fdkong Do you think this is something reasonable from PETSc? After changing the communicator for this error (see https://bitbucket.org/petsc/petsc/pull-requests/1835/change-the-communicator-for/diff), I get:
```
[2]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[2]PETSC ERROR: Null argument, when expecting valid pointer
[2]PETSC ERROR: No saved preallocation info 

[2]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[2]PETSC ERROR: Petsc Release Version 3.11.2, unknown 
[2]PETSC ERROR: [3]PETSC ERROR: --------------------- Error Message --------------------------------------------------------------
[3]PETSC ERROR: Null argument, when expecting valid pointer
[3]PETSC ERROR: No saved preallocation info 

[3]PETSC ERROR: See http://www.mcs.anl.gov/petsc/documentation/faq.html for trouble shooting.
[3]PETSC ERROR: Petsc Release Version 3.11.2, unknown 
[3]PETSC ERROR: ../../../moose_test-dbg on a arch-moose named Alexanders-Mac-Pro.local by lindad Mon Jul  1 14:22:08 2019
[3]PETSC ERROR: Configure options --download-hypre=1 --with-ssl=0 --with-debugging=no --with-pic=1 --with-shared-libraries=1 --with-cc=mpicc --with-cxx=mpicxx --with-fc=mpif90 --download-fblaslapack=1 --download-metis=1 --download-ptscotch=1 --download-parmetis=1 --download-superlu_dist=1 --download-mumps=1 --download-scalapack=1 --CC=mpicc --CXX=mpicxx --FC=mpif90 --F77=mpif77 --F90=mpif90 --CFLAGS="-fPIC -fopenmp" --CXXFLAGS="-fPIC -fopenmp" --FFLAGS="-fPIC -fopenmp" --FCFLAGS="-fPIC -fopenmp" --F90FLAGS="-fPIC -fopenmp" --F77FLAGS="-fPIC -fopenmp" --with-cxx-dialect=C++11 --with-fortran-bindings=0 --with-sowing=0
[3]PETSC ERROR: #1 MatResetPreallocation_SeqAIJ() line 3911 in /Users/lindad/projects/moose/petsc/src/mat/impls/aij/seq/aij.c
[3]PETSC ERROR: #2 MatResetPreallocation() line 847 in /Users/lindad/projects/moose/petsc/src/mat/interface/matrix.c
[3]PETSC ERROR: #3 MatResetPreallocation_MPIAIJ() line 2821 in /Users/lindad/projects/moose/petsc/src/mat/impls/aij/mpi/mpiaij.c
[3]PETSC ERROR: ../../../moose_test-dbg on a arch-moose named Alexanders-Mac-Pro.local by lindad Mon Jul  1 14:22:08 2019
#4 MatResetPreallocation() line 847 in /Users/lindad/projects/moose/petsc/src/mat/interface/matrix.c
[2]PETSC ERROR: Configure options --download-hypre=1 --with-ssl=0 --with-debugging=no --with-pic=1 --with-shared-libraries=1 --with-cc=mpicc --with-cxx=mpicxx --with-fc=mpif90 --download-fblaslapack=1 --download-metis=1 --download-ptscotch=1 --download-parmetis=1 --download-superlu_dist=1 --download-mumps=1 --download-scalapack=1 --CC=mpicc --CXX=mpicxx --FC=mpif90 --F77=mpif77 --F90=mpif90 --CFLAGS="-fPIC -fopenmp" --CXXFLAGS="-fPIC -fopenmp" --FFLAGS="-fPIC -fopenmp" --FCFLAGS="-fPIC -fopenmp" --F90FLAGS="-fPIC -fopenmp" --F77FLAGS="-fPIC -fopenmp" --with-cxx-dialect=C++11 --with-fortran-bindings=0 --with-sowing=0
[2]PETSC ERROR: #1 MatResetPreallocation_SeqAIJ() line 3911 in /Users/lindad/projects/moose/petsc/src/mat/impls/aij/seq/aij.c
[2]PETSC ERROR: #2 MatResetPreallocation() line 847 in /Users/lindad/projects/moose/petsc/src/mat/interface/matrix.c
[2]PETSC ERROR: #3 MatResetPreallocation_MPIAIJ() line 2821 in /Users/lindad/projects/moose/petsc/src/mat/impls/aij/mpi/mpiaij.c
[2]PETSC ERROR: #4 MatResetPreallocation() line 847 in /Users/lindad/projects/moose/petsc/src/mat/interface/matrix.c
```
when trying to run the above test case with 4 processes. This can be attributed to the fact that only two processes actually have dofs.

fdkong replied:
I see. I know what is going on. I will fix PETSc soon.  

But at this point, let us do some restrictions on moose if autoscaling is used. For example, we should have at least one dof per core. 

Sorry

lindsayad committed:
Check local dofs when doing auto scaling

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3176,6 +3176,14 @@ NonlinearSystemBase::computeScalingJacobian(NonlinearImplicitSystem & sys)
     auto & petsc_matrix = *static_cast<PetscMatrix<Real> *>(sys.matrix);
+
+    PetscInt m, n;
+    auto ierr = MatGetLocalSize(petsc_matrix.mat(), &m, &n);
```
ierr = MatGetLocalSize(petsc_matrix.mat(), &m, NULL);

If we do not care about m, just ignore it.

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3243,7 +3251,7 @@ NonlinearSystemBase::computeScalingJacobian(NonlinearImplicitSystem & sys)
 
     // Now it's essential that we reset the sparsity pattern of the matrix
-    auto ierr = MatResetPreallocation(petsc_matrix.mat());
+    ierr = MatResetPreallocation(petsc_matrix.mat());
```
I was wondering if it is possible to move `MatResetPreallocation`  and `MatGetLocalSize` to PetscSupport.C?


I am fine that  we have a lot of PETSc code in MOOSE, which will secure my job:-)  But other team members may try to keep things in one single place, and it is easier  to maintain.


Or add interfaces to `libMesh`. Up to you!

lindsayad replied:
Sigh, I will make the interfaces in libMesh. What a ruthless reviewer! 😄

fdkong replied:
I am trying to help you make your code even more beautiful.

lindsayad committed:
Use new PetscMatrix interfaces

fdkong commented:
@lindsayad libMesh is in for this PR?

lindsayad commented:
Not yet. Will do it on the plane today

> On Jul 10, 2019, at 11:41 AM, Fande Kong <notifications@github.com> wrote:
> 
> @lindsayad libMesh is in for this PR?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

lindsayad commented:
Now libMesh is in

lindsayad committed:
failed_solves.md -> FEProblemSolve.md for new design pattern

lindsayad commented:
@fdkong With your approval this should be ready to merge

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3180,11 +3180,17 @@ NonlinearSystemBase::computeScalingJacobian(NonlinearImplicitSystem & sys)
 
   if (dynamic_cast<PetscMatrix<Real> *>(sys.matrix))
   {
+#if !PETSC_VERSION_LESS_THAN(3, 8, 0)
```
Change it to 3.9.4.

fdkong commented on framework/src/systems/NonlinearSystemBase.C:
```diff
@@ -3250,6 +3256,15 @@ NonlinearSystemBase::computeScalingJacobian(NonlinearImplicitSystem & sys)
+    petsc_matrix.reset_preallocation();
+
+#else
+    mooseWarning("Automatic scaling requires a PETSc version of 3.8.0 or greater, so no automatic "
```
3.9.4

permcody replied:
So this warning breaks a few tests on the minimum compiler since it is running 3.7.6 still.

lindsayad commented:
It thought it was introduced in 3.9.0?

> On Jul 11, 2019, at 12:27 PM, Fande Kong <notifications@github.com> wrote:
> 
> @fdkong requested changes on this pull request.
> 
> In framework/src/systems/NonlinearSystemBase.C:
> 
> > @@ -3180,11 +3180,17 @@ NonlinearSystemBase::computeScalingJacobian(NonlinearImplicitSystem & sys)
>  
>    if (dynamic_cast<PetscMatrix<Real> *>(sys.matrix))
>    {
> +#if !PETSC_VERSION_LESS_THAN(3, 8, 0)
> Change it to 3.9.4.
> 
> In framework/src/systems/NonlinearSystemBase.C:
> 
> > @@ -3250,6 +3256,15 @@ NonlinearSystemBase::computeScalingJacobian(NonlinearImplicitSystem & sys)
>      applyScalingFactors(inverse_scaling_factors);
>      if (auto displaced_problem = _fe_problem.getDisplacedProblem().get())
>        displaced_problem->systemBaseNonlinear().applyScalingFactors(inverse_scaling_factors);
> +
> +    // Now it's essential that we reset the sparsity pattern of the matrix
> +    petsc_matrix.reset_preallocation();
> +
> +#else
> +    mooseWarning("Automatic scaling requires a PETSc version of 3.8.0 or greater, so no automatic "
> 3.9.4
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub, or mute the thread.

lindsayad commented:
Changed to 3.9.0 per git tags (and what we did in libMesh)

permcody commented:
So that function you removed was used by Rattlesnake. Anyone notice that or have a patch for that?

permcody commented:
I would have avoided merging this in since we probably should have fixed Rattlesnake, Mammoth, BlueCrab, Direwolf, etc first. No need to create pain for ourselves.
